sklearn/utils/sparsefuncs_fast.pyx

```diff
@@ -67,24 +67,38 @@ def csr_mean_variance_axis0(X):
         Feature-wise variances
 
     """
-    cdef unsigned int n_samples = X.shape[0]
-    cdef unsigned int n_features = X.shape[1]
-
-    cdef np.ndarray[DOUBLE, ndim=1, mode="c"] X_data
-    X_data = np.asarray(X.data, dtype=np.float64)     # might copy!
-    cdef np.ndarray[int, ndim=1] X_indices = X.indices
+    if X.dtype == np.int32:
```

Comment:
I think we're playing at the margins, but that keeping it as float64 (at least for int32) is safest. Most important of my comments is that other types remain supported.

Indeed, I think we've actually lost some precision in calculating the mean of float32s in float32 rather than float64, but in that case I think the difference is really marginal.